plutus-chain-index/app/Marconi.hs

```diff
@@ -174,50 +156,106 @@ getUtxoUpdate slot txs =
+  Coordinator <$> newBroadcastTChanIO
+              <*> newQSemN 0
+              <*> pure indexerCount
+
+datumWorker
+  :: Coordinator
+  -> TChan (ChainSyncEvent (BlockInMode CardanoMode))
+  -> FilePath
+  -> IO ()
+datumWorker Coordinator{_barrier} ch path = Datum.open path (Datum.Depth 2160) >>= innerLoop
+  where
+    innerLoop :: DatumIndex -> IO ()
+    innerLoop index = do
+      signalQSemN _barrier 1
+      event <- atomically $ readTChan ch
+      case event of
+        RollForward blk _ct ->
+          Ix.insert (getDatums blk) index >>= innerLoop
+        RollBackward cp _ct -> do
+          events <- Ix.getEvents (index ^. Ix.storage)
+          innerLoop $
+            fromMaybe index $ do
+              slot   <- chainPointToSlotNo cp
+              offset <- findIndex (any (\(s, _) -> s < slot)) events
+              Ix.rewind offset index
+
+utxoWorker
+  :: Coordinator
+  -> TChan (ChainSyncEvent (BlockInMode CardanoMode))
+  -> FilePath
+  -> IO ()
+utxoWorker Coordinator{_barrier} ch path = Utxo.open path (Utxo.Depth 2160) >>= innerLoop
+  where
+    innerLoop :: UtxoIndex -> IO ()
+    innerLoop index = do
+      signalQSemN _barrier 1
+      event <- atomically $ readTChan ch
+      case event of
+        RollForward (BlockInMode (Block (BlockHeader slotNo _ _) txs) _) _ct ->
+          Ix.insert (getUtxoUpdate slotNo txs) index >>= innerLoop
+        RollBackward cp _ct -> do
+          events <- Ix.getEvents (index ^. Ix.storage)
+          innerLoop $
+            fromMaybe index $ do
+              slot   <- chainPointToSlotNo cp
+              offset <- findIndex  (\u -> (u ^. Utxo.slotNo) < slot) events
+              Ix.rewind offset index
+
+combinedIndexer
+  :: Maybe FilePath
+  -> Maybe FilePath
   -> S.Stream (S.Of (ChainSyncEvent (BlockInMode CardanoMode))) IO r
   -> IO ()
-utxoIndexer path = S.foldM_ step initial finish
+combinedIndexer utxoPath datumPath = S.foldM_ step initial finish
   where
-    initial :: IO UtxoIndex
-    initial = Utxo.open path (Utxo.Depth 2160)
-
-    step :: UtxoIndex -> ChainSyncEvent (BlockInMode CardanoMode) -> IO UtxoIndex
-    step index (RollForward (BlockInMode (Block (BlockHeader slotNo _ _) txs) _) _ct) =
-      Ix.insert (getUtxoUpdate slotNo txs) index
-    step index (RollBackward cp _ct) = do
-      events <- Ix.getEvents (index ^. Ix.storage)
-      return $
-        fromMaybe index $ do
-          slot   <- chainPointToSlotNo cp
-          offset <- findIndex (\u -> (u ^. Utxo.slotNo) < slot) events
-          Ix.rewind offset index
-
-    finish :: UtxoIndex -> IO ()
-    finish _index = pure ()
+    initial :: IO Coordinator
+    initial = do
+      let indexerCount = length . catMaybes $ [utxoPath, datumPath]
+      coordinator <- initialCoordinator indexerCount
+      when (isJust datumPath) $ do
+        ch <- atomically . dupTChan $ _channel coordinator
+        void . forkIO . datumWorker coordinator ch $ fromJust datumPath
```

Comment:
Why not do the `dupTChan` inside the `datumWorker` instead of passing it as a parameter?

Reply:
Actually thinking about how this works, this is a possible multi-threading bug. Let me fix it first :)

Reply:
Ok. So there is no bug. 

Moving the channel creation in the function would mean I need to define an inner loop function which I preferred not to do.